total_scattering/autogrouping/autogrouping.py

```diff
@@ -0,0 +1,466 @@
+    print("Unique labels (clusters) = {} ({})".format(
+        len(unique_labels), unique_labels))
+    if len(unique_labels) == 1 and unique_labels[0] == -1:
+        # Print warning that the only labels found were noisy
+        skip_grouping = True
+        print("NOTE: only noisy clusters were found.. skipping "
+              "grouping generation!")
+
+    if method[1] == "ED":
+        if plots and plots["ED_Features"]:
+            fig, ax = plot_features(
+                clustering_input, diamond_peaks, fitparams,
+                labels, centroids)
+
+    if plots and plots["Grouping"]:
+        fig, ax = plt.subplots()
+        for i in range(len(wsindex)):
+            wsindex[i] = wksp.getDetector(i).getID()
+        ax.scatter(wsindex, labels + 1, c=labels)
+        ax.set_title("{}: {}".format(
+            grouping_method, diamond_file.split("/")[-1]))
+        ax.set_xlabel("pixel")
+        ax.set_ylabel("cluster/grouping")
+
+    if plots and plots["PCA"]:
+        pca = PCA(n_components=2)
+        pca_res = pca.fit_transform(clustering_input)
+        print("PCA variation of components: {}".format(
+            pca.explained_variance_ratio_))
+
+        fig, ax = plt.subplots()
+        ax.set_xlabel("PCA 1")
+        ax.set_ylabel("PCA 2")
+        ax.scatter(pca_res[:, 0], pca_res[:, 1], c=labels, alpha=0.5)
+        if method[0] == "KMEANS":
+            centroids_pca = pca.transform(centroids)
+            ax.scatter(centroids_pca[:, 0], centroids_pca[:, 1],
+                       marker='X', s=200, alpha=0.75,
+                       edgecolors='red', c=np.unique(labels))
+
+    # Export mask to file
+    if new_mask is not None:
+        print("Saving new mask to '{}'".format(output_mask))
+        # new mask is the union of original + newly masked items
+        if mask is not None:  # check in case no masking file was given
+            new_mask = np.union1d(mask, new_mask)
+        np.savetxt(output_mask, new_mask, fmt='%10i')
+
+    # Export grouping based on clustering result. Use the input
+    # workspace as the donor
+    if not skip_grouping:
+        print("Generating grouping file '{}'".format(output_file))
+        CreateGroupingWorkspace(InputWorkspace=wksp,
+                                OutputWorkspace="grouping")
+        grouping = mtd["grouping"]
+        for i in range(len(labels)):
+            det_id = wksp.getDetector(i).getID()
+            # Skip if label is -1 (DBSCAN labels these as noise)
+            if labels[i] == -1:
+                continue
+            # Shift by 1 since group 0 is unused
+            grouping.setY(det_id, [int(labels[i] + 1)])
+
+        SaveDetectorsGrouping(InputWorkspace=grouping,
+                              OutputFile=output_file)
+
+    if plots:
+        # block only if we have plotted something
+        nplot = sum(plot for plot in plots.values())
+        if nplot > 0:
+            plt.show(block=True)
+
+    return
```

Comment:
This function looks to have a really high cyclomatic complexity.
I am fine to leave as is for now but would like to eventually refactor into smaller functions to help readability (it is a bit terse right now)

Reply:
Agreed. But I do think we want to leave it for now and we will get back to it when people (including myself) are released from the PEMP required tasks which is due this month.